src/smt/optimization_solver.h

```diff
@@ -80,38 +90,350 @@ class Objective
-    SmtEngine* d_parent;
-    /** The objectives to optimize for **/
-    Objective d_activatedObjective;
-    /** A saved value of the objective from the last sat call. **/
-    Node d_savedValue;
-  };
+
+  /** Specify whether to use signed or unsigned comparison
+   * for BitVectors (only for BitVectors), this variable is defaulted to true
+   * **/
+  bool d_signed;
+};
+
+
+/**
+ * A solver for optimization queries.
+ *
+ * This class is responsible for responding to optmization queries. It
+ * spawns a subsolver SmtEngine that captures the parent assertions and
+ * implements a linear optimization loop. Supports activateObjective,
+ * checkOpt, and objectiveGetValue in that order.
+ */
+class OptimizationSolver
+{
+ public:
+  /** parent is the smt_solver that the user added their assertions to **/
+  OptimizationSolver(SmtEngine* parent);
+  ~OptimizationSolver();
+
+  /** Runs the optimization loop for the activated objective **/
+  OptResult checkOpt();
+  /** Activates an objective: will be optimized for
+   * Parameter is_signed specifies whether we should use signed/unsigned
+   * comparison for BitVectors (only effective for BitVectors) **/
+  void activateObj(const Node& obj,
+                   const int& type,
+                   bool bv_is_signed_compare = true);
+  /** Gets the value of the optimized objective after checkopt is called **/
+  Node objectiveGetValue();
+
+//  private:
+//   /** Returns the less than operator for the activated objective
+//    * if objective does not support comparison, return kind::NULL_EXPR **/
+//   Kind getLessThanOperatorForObjective();
+
+ private:
+  /** The parent SMT engine **/
+  SmtEngine* d_parent;
+  /** The objectives to optimize for **/
+  Objective d_activatedObjective;
+  /** A saved value of the objective from the last sat call. **/
+  Node d_savedValue;
+};
+
+/**
+ * Optimizer for individual datatype 
+ */
+struct OMTOptimizer 
+{
+  static std::unique_ptr<OMTOptimizer> getOptimizerForNode(CVC4::Node node, bool isSigned=false);
+  virtual ~OMTOptimizer() {}
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) = 0;
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) = 0;
+};
+
+/**
+ * The class is the optimizer implementation for individual types. 
+ */
+template <typename T>
+struct OMTOptimizerImpl : OMTOptimizer
+{ 
+  virtual ~OMTOptimizerImpl() {}
+  // for generic types, we couldn't optimize 
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) override {
+    return std::make_pair(OptResult::OPT_UNKNOWN, CVC4::Node());
+  }
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) override {
+    return std::make_pair(OptResult::OPT_UNKNOWN, CVC4::Node());
+  }
+};
+
+/**
+ * The goal is of type Integer 
+*/
+template <>
+struct OMTOptimizerImpl<Rational> : OMTOptimizer
+{
+public: 
```

Comment:
We typically do not implement inside header files.
https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#header-files